### PR TITLE
OKTA-560815 : Add support for engFastpassMultipleAccounts

### PR DESCRIPTION
**Pull request by leemchale-okta**
## Description:

This PR implements the OVFP behavior for when the engFastpassMultipleAccounts widget feature is enabled.

## PR Checklist

- [x] Have you verified the basic functionality for this change?
- [ ] Did you add tests, as appropriate, following our [Automated Test guidelines](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/2676497890/Automated+Testing+in+the+Signin+Widget)?
- [ ] Did you follow our [Security Best Practices](https://oktawiki.atlassian.net/wiki/display/eng/Security+Best+practices)?
- [ ] Did you verify the change by running [downstream monolith artifact](https://oktawiki.atlassian.net/wiki/spaces/eng/pages/102897979/Sign-in+Widget+Development#Sign-inWidgetDevelopment-Instructionstocreateandrunthedownstreamartifact(d16t))? (Provide link to build below)
- [ ] Does this PR include noticeable changes to the UI? (If yes, attach screenshots/video below)

### Issue:

- [OKTA-560815](https://oktainc.atlassian.net/browse/OKTA-560815)

### CODE REVIEW

File: src/v3/src/types/context.ts [hunk 1]
```diff
@@ -42,6 +42,8 @@ export type IWidgetContext = {
   loading: boolean;
   setLoading: StateUpdater<boolean>;
   setWidgetRendered: StateUpdater<boolean>;
+  loginHint?: string;
+  setLoginHint: StateUpdater<string | undefined>;
```
Comment (shawyu-okta) [inline src/v3/src/types/context.ts:46]: Is there a way to avoid adding these to the widget context? I'm not sure if we need state variables to hold the `loginHint`. I'm thinking it can potentially be pushed out to the transformers and we only use the `widgetContext` to get the feature flag info.

Comment (leemchale-okta) [reply src/v3/src/types/context.ts:46]: The problem was we needed a way for the loginHint/username to persist across views.  Since the `loginHint` is grabbed from the data in the identifier input field, we also need to be able to access it in the next flow with the Open Okta Verify button.

**Review by lesterchoi-okta — CHANGES_REQUESTED** [review]

File: test/testcafe/spec/DeviceChallengePollView_spec.js [hunk 2]
```diff
@@ -693,8 +690,9 @@ test
     await t.expect(deviceChallengePollPageObject.getFormTitle()).eql('Sign in with Okta FastPass');
 
     await t.expect(await deviceChallengePollPageObject.hasSpinner()).eql(true);
-    await t.expect(deviceChallengePollPageObject.getFooterCancelPollingLink().innerText).eql('Cancel and take me to sign in');
+    await t.expect(deviceChallengePollPageObject.getFooterCancelPollingLink().exists).eql(true);
 
+    await t.wait(5000); // wait for FASTPASS_FALLBACK_SPINNER_TIMEOUT
```
[the remark this answers is not in the document]
Comment (leemchale-okta) [reply test/testcafe/spec/DeviceChallengePollView_spec.js:695]: usually you can use the `waitFor` function to achieve that, but for some reason the `waitFor` function is not included in this testcafe context

File: src/v3/src/components/LaunchAuthenticatorButton/LaunchAuthenticatorButton.tsx [hunk 3]
```diff
@@ -39,13 +40,23 @@ const LaunchAuthenticatorButton: UISchemaElementComponent<{
...
+    userIdentifier,
+    setUserIdentifier,
+    data,
+    widgetProps: { features },
+  } = useWidgetContext();
 
   const handleClick: ClickHandler = async () => {
+    if (features?.engFastpassMultipleAccounts && data.identifier) {
+      // set userIdentifier in widget context to the current Username input field data
+      setUserIdentifier(encodeURIComponent(data.identifier as string));
```
[the remark this answers is not in the document]
Comment (shawyu-okta) [reply src/v3/src/components/LaunchAuthenticatorButton/LaunchAuthenticatorButton.tsx:53]: Just double-checking - is our ES6 polyfill configured to include this? Assuming we're using `core-js` then it should be included since it's part of most of the usual entry points: https://github.com/zloirock/core-js#url-and-urlsearchparams

File: src/v3/src/util/appendLoginHint.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+export const appendLoginHint = (deviceChallengeUrl = '', loginHint = ''): string => {
+  if (deviceChallengeUrl && loginHint) {
+    return `${deviceChallengeUrl}&login_hint=${loginHint}`;
+  }
+  return deviceChallengeUrl;
+};
```
[the remark this answers is not in the document]
Comment (shawyu-okta) [reply src/v3/src/util/appendLoginHint.ts:18]: Agree on both counts, we can rename the fn and also improve on the original implementation in v2 by using `URL` and `URLSearchParams` while we're at it

`URL`: https://developer.mozilla.org/en-US/docs/Web/API/URL/URL
`URL.searchParams` is a `URLSearchParams` object: https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams

File: src/v3/src/components/Widget/index.tsx [hunk 5]
```diff
@@ -282,6 +283,8 @@ export const Widget: FunctionComponent<WidgetProps> = (widgetProps) => {
       loading,
       setLoading,
       setWidgetRendered,
+      userIdentifier,
+      setUserIdentifier,
```
Comment (shawyu-okta) [inline src/v3/src/components/Widget/index.tsx:287]: We need to make sure this value is cleared in the case that a user clears the username field (should go back to `null`/`undefined` OR we need to check for empty strings)

Also if a user backs out of a flow by clicking the back button in the widget and the widget shows a blank username field again, we need to ensure the value is cleared as well.

Finally, what happens if the username is prefilled via widgetConfig or rememberMe functionality?

File: src/v3/src/components/Widget/index.tsx [hunk 6]
```diff
@@ -96,6 +96,7 @@ export const Widget: FunctionComponent<WidgetProps> = (widgetProps) => {
   const [loading, setLoading] = useState<boolean>(false);
   const [widgetRendered, setWidgetRendered] = useState<boolean>(false);
   const brandedTheme = mapMuiThemeFromBrand(brandColors);
+  const [userIdentifier, setUserIdentifier] = useState<string | undefined>(undefined);
```
Comment (shawyu-okta) [inline src/v3/src/components/Widget/index.tsx:99]: To go along with the comment below, if we still want to allow non-string values `null | undefined` I favor using `null` when the value is intentionally empty.

Another way to go is _only_ allow `string` and check for empty strings in the component code before appending the parameters to the URL.

File: src/v3/src/util/appendLoginHint.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (leemchale-okta) [reply src/v3/src/util/appendLoginHint.ts:18]: I found a problem in that this breaks the spec test since the spec test uses `okta-verify.html` as the dummy url.  The `URL` object constructor does not accept that as a valid url and throws an error:
image

How do we want to handle this do we want to change the test mock file?

File: src/v3/src/components/Widget/index.tsx (hunk 5 above)
Comment (leemchale-okta) [reply src/v3/src/components/Widget/index.tsx:287]: if the username is prefilled via widgetConfig or rememberMe it will still work properly since it sets the `userIdentifier` value to whatever is in the username field only after the launch authenticator button is pressed.  The username field data is populated before the authenticator button is pressed.  

Since the `userIdentifier` value is set only after the launch auth button is pressed, we shouldnt have any issue of possibly having the wrong username set in it or the user clearing the field.  So it shouldnt be a problem if someone backs out of the flow, but technically the old username would still be leftover in `userIdentifier` until the launch auth button is pressed again which may be a security concern.  I think a simple solution to this would be to clear the `userIdentifier` as soon as the identify flow loads.

Comment (shawyu-okta) [reply src/v3/src/components/Widget/index.tsx:287]: > if the username is prefilled via widgetConfig or rememberMe it will still work properly since it sets the userIdentifier value to whatever is in the username field only after the launch authenticator button is pressed. The username field data is populated before the authenticator button is pressed.

👍 

> So it shouldnt be a problem if someone backs out of the flow, but technically the old username would still be leftover in userIdentifier until the launch auth button is pressed again which may be a security concern. I think a simple solution to this would be to clear the userIdentifier as soon as the identify flow loads.

Cool, I think as a part of good hygiene we should clear out these widget context values properly in this case even if it's not affecting the widget behavior in the current use-case.

File: src/v3/src/types/context.ts [hunk 7]
```diff
@@ -42,6 +42,8 @@ export type IWidgetContext = {
   loading: boolean;
   setLoading: StateUpdater<boolean>;
   setWidgetRendered: StateUpdater<boolean>;
+  userIdentifier?: string;
+  setUserIdentifier: StateUpdater<string | undefined>;
```
Comment (glenfannin-okta) [inline src/v3/src/types/context.ts:46]: Any thoughts on this name? In v2 it is set as `login_hint`, we don't have to mimic this exactly, but userIdentifier is a bit misleading because it may not actually be the real identifier, and it's only used in specific scenarios. I just don't want this to be confused later down the road with something it isn't.

Comment (leemchale-okta) [reply src/v3/src/types/context.ts]: Thats a good point since we are grabbing the username from the input field, it may not necessarily be a valid username.  I believe it was changed because we wanted to make the name more generic.  @shawyu-okta what do you think?

File: src/v3/src/components/LaunchAuthenticatorButton/LaunchAuthenticatorButton.tsx [hunk 8]
```diff
@@ -39,13 +40,24 @@ const LaunchAuthenticatorButton: UISchemaElementComponent<{
...
   const focusRef = useAutoFocus<HTMLButtonElement>(focus);
+  const {
+    userIdentifier,
+    setUserIdentifier,
+    data,
+    widgetProps: { features },
+  } = useWidgetContext();
 
   const handleClick: ClickHandler = async () => {
+    if (features?.engFastpassMultipleAccounts && data.identifier) {
```
Comment (lesterchoi-okta) [inline src/v3/src/components/LaunchAuthenticatorButton/LaunchAuthenticatorButton.tsx:51]:
```suggestion
    if (data.identifier) {
```

Comment (lesterchoi-okta) [reply src/v3/src/components/LaunchAuthenticatorButton/LaunchAuthenticatorButton.tsx:51]: Per latest comment in Jira, we are removing this flag check

Comment (lesterchoi-okta) [reply src/v3/src/components/LaunchAuthenticatorButton/LaunchAuthenticatorButton.tsx:51]: https://github.com/okta/okta-signin-widget/pull/3034

File: src/v3/src/types/context.ts (hunk 7 above)
Comment (lesterchoi-okta) [reply src/v3/src/types/context.ts]: Anything wrong with the obvious alternative?

```
  loginHint?: string;
  setLoginHint: StateUpdater<string | undefined>;
```

Comment (shawyu-okta) [reply src/v3/src/types/context.ts]: Yeah I had @leemchale-okta change it to `userIdentifier` because I wanted to avoid a very specific global context value if possible, but based on the comments above I don't have an issue reverting to `loginHint`

File: src/v3/src/components/LaunchAuthenticatorButton/LaunchAuthenticatorButton.tsx (hunk 8 above)
Comment (shawyu-okta) [reply src/v3/src/components/LaunchAuthenticatorButton/LaunchAuthenticatorButton.tsx:51]: Let's remove the check here and update the tests in the same fashion as in #3034. It will cause some conflicts on rebase/merge to the trunk but they shouldn't be difficult to resolve.

File: src/v3/src/hooks/useOnSubmit.ts [hunk 9]
```diff
@@ -168,6 +169,10 @@ export const useOnSubmit = (): (options: OnSubmitHandlerOptions) => Promise<void
           i18n: { key: 'oform.errorbanner.title' },
         } as IdxMessage);
       }
+      // clear the userIdentifier value when returning to the identify flow
+      if (newTransaction.nextStep?.name === IDX_STEP.IDENTIFY) {
+        setUserIdentifier(null);
+      }
```
Comment (shawyu-okta) [inline src/v3/src/hooks/useOnSubmit.ts:175]: It might be better to put this setter somewhere where we react to the transaction changing, rather than where we are setting the changed transaction, if that makes sense. I view this hook as a place where we are updating the transaction value, whereas the bit about clearing the userIdentifier value is a reaction to when that changes.

We have some effect hooks that are in `Widget/index` that might be more appropriate since they are run whenever transaction changes.

File: src/v3/src/components/OpenOktaVerifyFPButton/OpenOktaVerifyFPButton.tsx [hunk 10]
```diff
@@ -53,6 +54,9 @@ const OpenOktaVerifyFPButton: UISchemaElementComponent<{
   } = uischema;
   const [key, setKey] = useState<number>(0);
   const label = getTranslation(translations, 'label');
+  const { userIdentifier } = useWidgetContext();
+  const loginHintQueryParam = userIdentifier ? { login_hint: userIdentifier } : undefined;
```
Comment (shawyu-okta) [inline src/v3/src/components/OpenOktaVerifyFPButton/OpenOktaVerifyFPButton.tsx:58]: nit: consider `null`
```suggestion
  const loginHintQueryParam = userIdentifier ? { login_hint: userIdentifier } : null;
```

File: src/v3/src/types/context.ts (hunk 7 above)
Comment (glenfannin-okta) [reply src/v3/src/types/context.ts]: Ahh, that makes sense, my only concern was if a home team eng comes across this, they may think this is some quick way to access the user's identifier. Granted that could be dispelled rather quickly with a codebase search, but just wanted to get your opinion on this. I am good with either way 🙏

File: src/v3/src/util/setUrlQueryParams.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+export const setUrlQueryParams = (
+  url: URL | string,
+  params: Record<string, string> = {},
+) => url + '?' + new URLSearchParams(params).toString();
```
Comment (shawyu-okta) [inline src/v3/src/util/setUrlQueryParams.ts:16]: https://github.com/okta/okta-signin-widget/pull/3031#discussion_r1087255593
> I found a problem in that this breaks the spec test since the spec test uses okta-verify.html as the dummy url. The URL object constructor does not accept that as a valid url and throws an error:

There's a corner case here where if the `url` already has the `?` in it then this might return an invalid URL, but in practice I don't think we'll run into it.

Using the `URL` class should mitigate this and you might be able to use the `URL(url, base)` overload of the constructor and pass the widget `baseUrl` to it. We have a utility in `settingsUtils` called `getBaseUrl(widgetProps)`, although it looks like it can still return `undefined`.

**Review by shawyu-okta — APPROVED** [review]
Only blocking comments to address are to remove the FF / update tests, changing the context value for identifier to be named hint, and the location of where we clear out the identifier/hint value.

File: src/v3/src/components/LaunchAuthenticatorButton/LaunchAuthenticatorButton.tsx (hunk 8 above)
Comment (leemchale-okta) [reply src/v3/src/components/LaunchAuthenticatorButton/LaunchAuthenticatorButton.tsx:51]: 👍

File: src/v3/src/hooks/useOnSubmit.ts (hunk 9 above)
Comment (leemchale-okta) [reply src/v3/src/hooks/useOnSubmit.ts:175]: Okay yeah I see what youre saying.  The `useOnSubmit` seemed like a reliable place to put it but the `Widget/index` is probably more appropriate.

File: src/v3/src/types/context.ts (hunk 7 above)
Comment (leemchale-okta) [reply src/v3/src/types/context.ts]: okay makes sense I will change it to `  loginHint?: string;
  setLoginHint: StateUpdater<string | undefined>;`

File: src/v3/src/components/OpenOktaVerifyFPButton/OpenOktaVerifyFPButton.tsx [hunk 12]
```diff
@@ -53,6 +56,10 @@ const OpenOktaVerifyFPButton: UISchemaElementComponent<{
   } = uischema;
   const [key, setKey] = useState<number>(0);
   const label = getTranslation(translations, 'label');
+  const { loginHint, widgetProps } = useWidgetContext();
+  const urlObj = new URL(href as string, getBaseUrl(widgetProps));
```
Comment (lesterchoi-okta) [inline src/v3/src/components/OpenOktaVerifyFPButton/OpenOktaVerifyFPButton.tsx:60]: 👍

**Review by lesterchoi-okta — APPROVED** [review]
